core/src/main/java/org/apache/iceberg/encryption/KeyMetadataDecoder.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.message.BadHeaderException;
+import org.apache.avro.message.MessageDecoder;
+import org.apache.avro.message.MissingSchemaException;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.avro.ProjectionDatumReader;
+import org.apache.iceberg.data.avro.DataReader;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.MapMaker;
+
+public class KeyMetadataDecoder<D> extends MessageDecoder.BaseDecoder<D> {
+  private final org.apache.iceberg.Schema readSchema;
+  private final Map<Byte, RawDecoder<D>> decoders = new MapMaker().makeMap();
+  private final byte[] versionBuffer;
+
+  /**
+   * Creates a new decoder that constructs datum instances described by an {@link
+   * org.apache.iceberg.Schema Iceberg schema}.
+   *
+   * <p>The {@code readSchema} is as used the expected schema (read schema). Datum instances created
+   * by this class will are described by the expected schema.
+   *
+   * @param readSchema the schema used to construct datum instances
+   */
+  public KeyMetadataDecoder(byte currentSchemaVersion, org.apache.iceberg.Schema readSchema) {
```

Comment:
Why does this pass the version in? Since we are defining versions, this could just have the schemas embedded in the class already and use a static map of schemas.

Reply:
Sounds good.